File: sdks/javascript/violationsentinel.ts
/**
 * ViolationSentinel JavaScript/TypeScript SDK
 * Official JavaScript client for the ViolationSentinel API
 * 
 * Installation:
 *   npm install @violationsentinel/sdk
 * 
 * Usage:
 *   import { ViolationSentinelClient } from '@violationsentinel/sdk';
 *   
 *   const client = new ViolationSentinelClient({
 *     apiKey: 'your-api-key',
 *     tenantId: 'your-tenant-id'
 *   });
 *   
 *   // List properties
 *   const properties = await client.properties.list();
 *   
 *   // Get violations
 *   const violations = await client.violations.list({ propertyId: 'prop-123' });
 */

interface ClientConfig {
  apiKey: string;
  tenantId: string;
  baseUrl?: string;
  timeout?: number;
}

interface Property {
  id: string;
  name: string;
  bbl: string;
  address?: string;
  year_built?: number;
  units?: number;
  is_monitored: boolean;
  tenant_id: string;
  created_at: string;
}

interface Violation {
  id: string;
  property_id: string;
  source: string;
  external_id: string;
  description?: string;
  violation_class?: string;
  issued_date?: string;
  resolved_date?: string;
  is_resolved: boolean;
  risk_score: number;
  confidence_score: number;
  created_at: string;
}

interface ScanRequest {
  property_ids?: string[];
  scan_all?: boolean;
  sources?: string[];
}

interface Report {
  report_id: string;
  status: string;
  download_url?: string;
  generated_at: string;
}

interface DashboardMetrics {
  total_properties: number;
  monitored_properties: number;
  total_violations: number;
  high_risk_properties: number;
  violations_last_30_days: number;
  avg_risk_score: number;
}

class ViolationSentinelError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ViolationSentinelError';
  }
}

class AuthenticationError extends ViolationSentinelError {
  constructor(message: string = 'Authentication failed') {
    super(message);
    this.name = 'AuthenticationError';
  }
}

class RateLimitError extends ViolationSentinelError {
  constructor(retryAfter: number) {
    super(`Rate limit exceeded. Retry after ${retryAfter} seconds`);
    this.name = 'RateLimitError';
  }
}

class APIError extends ViolationSentinelError {
  constructor(message: string, public statusCode: number) {
    super(message);
    this.name = 'APIError';
  }
}

export class ViolationSentinelClient {
  private apiKey: string;
  private tenantId: string;
  private baseUrl: string;
  private timeout: number;

  public properties: PropertiesAPI;
  public violations: ViolationsAPI;
  public reports: ReportsAPI;
  public webhooks: WebhooksAPI;
  public analytics: AnalyticsAPI;

  constructor(config: ClientConfig) {
    this.apiKey = config.apiKey;
    this.tenantId = config.tenantId;
    this.baseUrl = (config.baseUrl || 'https://api.violationsentinel.com').replace(/\/$/, '');
    this.timeout = config.timeout || 30000;

    this.properties = new PropertiesAPI(this);
    this.violations = new ViolationsAPI(this);
    this.reports = new ReportsAPI(this);
    this.webhooks = new WebhooksAPI(this);
    this.analytics = new AnalyticsAPI(this);
  }

  async request<T>(
    method: string,
    endpoint: string,
    options: {
      params?: Record<string, any>;
      data?: any;
    } = {}
  ): Promise<T> {
    const url = new URL(`${this.baseUrl}/api/v1${endpoint}`);
    
    // Add query parameters
    if (options.params) {
      Object.entries(options.params).forEach(([key, value]) => {
        if (value !== undefined && value !== null) {
          url.searchParams.append(key, String(value));
        }
      });
    }

    const headers: Record<string, string> = {
      'X-API-Key': this.apiKey,
      'X-Tenant-ID': this.tenantId,
      'Content-Type': 'application/json',
      'User-Agent': 'ViolationSentinel-JS-SDK/1.0.0'
    };

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeout);

    try {
      const response = await fetch(url.toString(), {
        method,
        headers,
        body: options.data ? JSON.stringify(options.data) : undefined,
        signal: controller.signal
      });

      clearTimeout(timeoutId);

      // Handle rate limiting
      if (response.status === 429) {
        const retryAfter = parseInt(response.headers.get('Retry-After') || '60');
        throw new RateLimitError(retryAfter);
      }

      // Handle authentication errors
      if (response.status === 401) {
        throw new AuthenticationError('Invalid API key or tenant ID');
      }

      // Handle other errors
      if (!response.ok) {
        const errorData = await response.json().catch(() => ({}));
        throw new APIError(
          errorData.detail || 'Unknown error',
          response.status
        );
      }

      // Handle no-content responses
      if (response.status === 204) {
        return undefined as T;
      }

      return await response.json();
    } catch (error) {
      if (error instanceof ViolationSentinelError) {
        throw error;
      }
      throw new ViolationSentinelError(`Request failed: ${error.message}`);
    }
  }
}

class PropertiesAPI {
  constructor(private client: ViolationSentinelClient) {}

  async list(skip: number = 0, limit: number = 100): Promise<Property[]> {
    return this.client.request<Property[]>('GET', '/properties', {
      params: { skip, limit }
    });
  }

  async get(propertyId: string): Promise<Property> {
    return this.client.request<Property>('GET', `/properties/${propertyId}`);
  }

  async create(data: {
    name: string;
    bbl: string;
    address?: string;
    year_built?: number;
    units?: number;
  }): Promise<Property> {
    return this.client.request<Property>('POST', '/properties', { data });
  }

  async update(propertyId: string, data: Partial<Property>): Promise<Property> {
    return this.client.request<Property>('PUT', `/properties/${propertyId}`, { data });
  }

  async delete(propertyId: string): Promise<void> {
    return this.client.request<void>('DELETE', `/properties/${propertyId}`);
  }
}

class ViolationsAPI {
  constructor(private client: ViolationSentinelClient) {}

  async list(filters: {
    property_id?: string;
    source?: string;
    is_resolved?: boolean;
    skip?: number;
    limit?: number;
  } = {}): Promise<Violation[]> {
    return this.client.request<Violation[]>('GET', '/violations', {
      params: filters
    });
  }

  async get(violationId: string): Promise<Violation> {
    return this.client.request<Violation>('GET', `/violations/${violationId}`);
  }

  async scan(request: ScanRequest): Promise<any> {
    return this.client.request('POST', '/violations/scan', { data: request });
  }
}

class ReportsAPI {
  constructor(private client: ViolationSentinelClient) {}

  async generate(data: {
    property_ids: string[];
    start_date?: string;
    end_date?: string;
    include_resolved?: boolean;
    format?: string;
  }): Promise<Report> {
    return this.client.request<Report>('POST', '/reports', { data });
  }

  async getStatus(reportId: string): Promise<Report> {
    return this.client.request<Report>('GET', `/reports/${reportId}`);
  }
}

class WebhooksAPI {
  constructor(private client: ViolationSentinelClient) {}

  async list(): Promise<any[]> {
    return this.client.request<any[]>('GET', '/webhooks');
  }

  async create(data: {
    url: string;
    events: string[];
    is_active?: boolean;
  }): Promise<any> {
    return this.client.request('POST', '/webhooks', { data });
  }

  async delete(webhookId: string): Promise<void> {
    return this.client.request<void>('DELETE', `/webhooks/${webhookId}`);
  }
}

class AnalyticsAPI {
  constructor(private client: ViolationSentinelClient) {}

  async dashboard(): Promise<DashboardMetrics> {
    return this.client.request<DashboardMetrics>('GET', '/analytics/dashboard');
  }

  async violationStats(): Promise<any> {
    return this.client.request('GET', '/analytics/violations/stats');
  }
}

// Example usage
export default async function example() {
  const client = new ViolationSentinelClient({
    apiKey: 'your-api-key',
    tenantId: 'your-tenant-id'
  });

  // List properties
  const properties = await client.properties.list();
  console.log(`Found ${properties.length} properties`);

  // Create property
  const newProperty = await client.properties.create({
    name: '123 Main Street',
    bbl: '1012650001',
    address: '123 Main St, New York, NY 10001',
    year_built: 1950,
    units: 24
  });
  console.log(`Created property: ${newProperty.id}`);

  // Get violations
  const violations = await client.violations.list({
    property_id: newProperty.id,
    is_resolved: false
  });
  console.log(`Found ${violations.length} open violations`);

  // Trigger scan
  const scanResult = await client.violations.scan({
    property_ids: [newProperty.id],
    sources: ['DOB', 'HPD', '311']
  });
  console.log(`Scan started: ${scanResult.scan_id}`);

  // Get dashboard metrics
  const metrics = await client.analytics.dashboard();
  console.log(`Total violations: ${metrics.total_violations}`);
}

File: sdks/python/violationsentinel.py
"""
ViolationSentinel Python SDK
Official Python client for the ViolationSentinel API

Installation:
    pip install violationsentinel

Usage:
    from violationsentinel import ViolationSentinelClient
    
    client = ViolationSentinelClient(
        api_key="your-api-key",
        tenant_id="your-tenant-id"
    )
    
    # List properties
    properties = client.properties.list()
    
    # Get property violations
    violations = client.violations.list(property_id="prop-123")
"""

import requests
from typing import List, Dict, Optional, Any
from datetime import datetime
import json


class ViolationSentinelError(Exception):
    """Base exception for ViolationSentinel SDK"""
    pass


class AuthenticationError(ViolationSentinelError):
    """Raised when authentication fails"""
    pass


class APIError(ViolationSentinelError):
    """Raised when API returns an error"""
    pass


class RateLimitError(ViolationSentinelError):
    """Raised when rate limit is exceeded"""
    pass


class ViolationSentinelClient:
    """Main client for ViolationSentinel API"""
    
    def __init__(
        self,
        api_key: str,
        tenant_id: str,
        base_url: str = "https://api.violationsentinel.com",
        timeout: int = 30
    ):
        """
        Initialize ViolationSentinel client
        
        Args:
            api_key: Your API key
            tenant_id: Your tenant ID
            base_url: Base URL for API (default: production)
            timeout: Request timeout in seconds
        """
        self.api_key = api_key
        self.tenant_id = tenant_id
        self.base_url = base_url.rstrip('/')
        self.timeout = timeout
        
        self.properties = PropertiesAPI(self)
        self.violations = ViolationsAPI(self)
        self.reports = ReportsAPI(self)
        self.webhooks = WebhooksAPI(self)
        self.analytics = AnalyticsAPI(self)
    
    def _request(
        self,
        method: str,
        endpoint: str,
        params: Optional[Dict] = None,
        data: Optional[Dict] = None,
        json_data: Optional[Dict] = None
    ) -> Dict[str, Any]:
        """Make HTTP request to API"""
        url = f"{self.base_url}/api/v1{endpoint}"
        
        headers = {
            "X-API-Key": self.api_key,
            "X-Tenant-ID": self.tenant_id,
            "Content-Type": "application/json",
            "User-Agent": "ViolationSentinel-Python-SDK/1.0.0"
        }
        
        try:
            response = requests.request(
                method=method,
                url=url,
                headers=headers,
                params=params,
                data=data,
                json=json_data,
                timeout=self.timeout
            )
            
            # Handle rate limiting
            if response.status_code == 429:
                retry_after = response.headers.get('Retry-After', '60')
                raise RateLimitError(
                    f"Rate limit exceeded. Retry after {retry_after} seconds"
                )
            
            # Handle authentication errors
            if response.status_code == 401:
                raise AuthenticationError("Invalid API key or tenant ID")
            
            # Handle other errors
            if response.status_code >= 400:
                error_msg = response.json().get('detail', 'Unknown error')
                raise APIError(f"API error: {error_msg} (status: {response.status_code})")
            
            return response.json()
            
        except requests.RequestException as e:
            raise ViolationSentinelError(f"Request failed: {str(e)}")


class PropertiesAPI:
    """Properties API client"""
    
    def __init__(self, client: ViolationSentinelClient):
        self.client = client
    
    def list(self, skip: int = 0, limit: int = 100) -> List[Dict]:
        """List all properties"""
        return self.client._request(
            "GET",
            "/properties",
            params={"skip": skip, "limit": limit}
        )
    
    def get(self, property_id: str) -> Dict:
        """Get property by ID"""
        return self.client._request("GET", f"/properties/{property_id}")
    
    def create(
        self,
        name: str,
        bbl: str,
        address: Optional[str] = None,
        year_built: Optional[int] = None,
        units: Optional[int] = None
    ) -> Dict:
        """Create new property"""
        data = {
            "name": name,
            "bbl": bbl,
            "address": address,
            "year_built": year_built,
            "units": units
        }
        return self.client._request("POST", "/properties", json_data=data)
    
    def update(self, property_id: str, **kwargs) -> Dict:
        """Update property"""
        return self.client._request(
            "PUT",
            f"/properties/{property_id}",
            json_data=kwargs
        )
    
    def delete(self, property_id: str) -> None:
        """Delete property"""
        self.client._request("DELETE", f"/properties/{property_id}")


class ViolationsAPI:
    """Violations API client"""
    
    def __init__(self, client: ViolationSentinelClient):
        self.client = client
    
    def list(
        self,
        property_id: Optional[str] = None,
        source: Optional[str] = None,
        is_resolved: Optional[bool] = None,
        skip: int = 0,
        limit: int = 100
    ) -> List[Dict]:
        """List violations with filters"""
        params = {"skip": skip, "limit": limit}
        if property_id:
            params["property_id"] = property_id
        if source:
            params["source"] = source
        if is_resolved is not None:
            params["is_resolved"] = is_resolved
        
        return self.client._request("GET", "/violations", params=params)
    
    def get(self, violation_id: str) -> Dict:
        """Get violation by ID"""
        return self.client._request("GET", f"/violations/{violation_id}")
    
    def scan(
        self,
        property_ids: Optional[List[str]] = None,
        scan_all: bool = False,
        sources: List[str] = None
    ) -> Dict:
        """Trigger violation scan"""
        data = {
            "property_ids": property_ids,
            "scan_all": scan_all,
            "sources": sources or ["DOB", "HPD", "311"]
        }
        return self.client._request("POST", "/violations/scan", json_data=data)


class ReportsAPI:
    """Reports API client"""
    
    def __init__(self, client: ViolationSentinelClient):
        self.client = client
    
    def generate(
        self,
        property_ids: List[str],
        start_date: Optional[datetime] = None,
        end_date: Optional[datetime] = None,
        include_resolved: bool = False,
        format: str = "json"
    ) -> Dict:
        """Generate compliance report"""
        data = {
            "property_ids": property_ids,
            "start_date": start_date.isoformat() if start_date else None,
            "end_date": end_date.isoformat() if end_date else None,
            "include_resolved": include_resolved,
            "format": format
        }
        return self.client._request("POST", "/reports", json_data=data)
    
    def get_status(self, report_id: str) -> Dict:
        """Get report generation status"""
        return self.client._request("GET", f"/reports/{report_id}")


class WebhooksAPI:
    """Webhooks API client"""
    
    def __init__(self, client: ViolationSentinelClient):
        self.client = client
    
    def list(self) -> List[Dict]:
        """List all webhooks"""
        return self.client._request("GET", "/webhooks")
    
    def create(
        self,
        url: str,
        events: List[str],
        is_active: bool = True
    ) -> Dict:
        """Create webhook subscription"""
        data = {
            "url": url,
            "events": events,
            "is_active": is_active
        }
        return self.client._request("POST", "/webhooks", json_data=data)
    
    def delete(self, webhook_id: str) -> None:
        """Delete webhook"""
        self.client._request("DELETE", f"/webhooks/{webhook_id}")


class AnalyticsAPI:
    """Analytics API client"""
    
    def __init__(self, client: ViolationSentinelClient):
        self.client = client
    
    def dashboard(self) -> Dict:
        """Get dashboard metrics"""
        return self.client._request("GET", "/analytics/dashboard")
    
    def violation_stats(self) -> Dict:
        """Get violation statistics"""
        return self.client._request("GET", "/analytics/violations/stats")


# Example usage
if __name__ == "__main__":
    # Initialize client
    client = ViolationSentinelClient(
        api_key="your-api-key",
        tenant_id="your-tenant-id"
    )
    
    # List properties
    properties = client.properties.list()
    print(f"Found {len(properties)} properties")
    
    # Create property
    new_property = client.properties.create(
        name="123 Main Street",
        bbl="1012650001",
        address="123 Main St, New York, NY 10001",
        year_built=1950,
        units=24
    )
    print(f"Created property: {new_property['id']}")
    
    # Get violations
    violations = client.violations.list(
        property_id=new_property['id'],
        is_resolved=False
    )
    print(f"Found {len(violations)} open violations")
    
    # Trigger scan
    scan_result = client.violations.scan(
        property_ids=[new_property['id']],
        sources=["DOB", "HPD", "311"]
    )
    print(f"Scan started: {scan_result['scan_id']}")
    
    # Get dashboard metrics
    metrics = client.analytics.dashboard()
    print(f"Total violations: {metrics['total_violations']}")
    
    # Generate report
    report = client.reports.generate(
        property_ids=[new_property['id']],
        format="pdf"
    )
    print(f"Report generating: {report['report_id']}")
